### PR TITLE
hls: Add BaseHLS utility function tests

### PR DESCRIPTION
**Pull request by sjalander**

### CODE REVIEW

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+          jlm::rvsdg::FunctionType::Create(
+              { valueType, valueType, valueType },
+              { valueType, valueType }),
+          "f",
+          jlm::llvm::Linkage::externalLinkage));
+
+  // Finalize the lambda to create results
+  auto f =
+      lambda->finalize({ lambda->GetFunctionArguments()[0], lambda->GetFunctionArguments()[1] });
+  (void)f;
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: just avoid declaring f in the first place if it is not used.

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+
+  // Act & Assert - use subregion() to get results
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[0]), "a0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[1]), "a1");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[2]), "a2");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->subregion()->result(0)), "r0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->subregion()->result(1)), "r1");
+}
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Be consistent: Use lambda->GetFunctionResults()[]

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+      rvsdg.GetRootRegion(),
+      jlm::llvm::LlvmLambdaOperation::Create(
+          jlm::rvsdg::FunctionType::Create({ valueType }, { valueType, valueType }),
+          "f",
+          jlm::llvm::Linkage::externalLinkage));
+
+  // Finalize the lambda to create results
+  auto f =
+      lambda->finalize({ lambda->GetFunctionArguments()[0], lambda->GetFunctionArguments()[0] });
+  (void)f;
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: same here

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+
+  // Act
+  auto regResults = TestableBaseHLS().get_reg_results(*lambda);
+
+  // Assert
+  EXPECT_EQ(regResults.size(), 2);
+  EXPECT_EQ(regResults[0], lambda->subregion()->result(0));
+  EXPECT_EQ(regResults[1], lambda->subregion()->result(1));
+}
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: GetFunctionResults()

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+  auto lambda = jlm::rvsdg::LambdaNode::Create(
+      rvsdg.GetRootRegion(),
+      jlm::llvm::LlvmLambdaOperation::Create(
+          jlm::rvsdg::FunctionType::Create({ valueType }, { valueType }),
+          "f",
+          jlm::llvm::Linkage::externalLinkage));
+
+  // Finalize the lambda to create results
+  auto f = lambda->finalize({ lambda->GetFunctionArguments()[0] });
+  (void)f;
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: same here

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+TEST(BaseHlsTests, TestJlmSize)
+{
+  // Test with bit type
+  auto bitType = jlm::rvsdg::BitType::Create(32);
+  EXPECT_EQ(TestableBaseHLS().JlmSize(bitType.get()), 32);
+
+  // Test with pointer type
+  auto ptrType = jlm::llvm::PointerType::Create();
+  EXPECT_EQ(TestableBaseHLS().JlmSize(ptrType.get()), 64); // Assuming 64-bit system
+}
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: What about other types?

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+  auto lambda = jlm::rvsdg::LambdaNode::Create(
+      rvsdg.GetRootRegion(),
+      jlm::llvm::LlvmLambdaOperation::Create(
+          jlm::rvsdg::FunctionType::Create({ bitType }, { bitType }),
+          "f",
+          jlm::llvm::Linkage::externalLinkage));
+
+  // Finalize the lambda to create results
+  auto f = lambda->finalize({ lambda->GetFunctionArguments()[0] });
+  (void)f;
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: same here

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+
+  // Act & Assert
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[0]), "a0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->subregion()->result(0)), "r0");
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: GetFunctionResults()

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+}
+
+// Test memory request extraction with BundleType results
+TEST(BaseHlsTests, TestMemReqsWithBundleType)
+{
+  // Note: This test demonstrates that get_mem_reqs() checks for BundleType results.
+  // However, creating actual BundleType results requires finalize() to be called,
+  // which needs inputs. Since we can't create a lambda with only BundleType outputs
+  // without inputs (no values to finalize), this test verifies the behavior when
+  // no memory requests exist in a typical lambda structure.
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: finalize does not need inputs. It can be invoked with {}

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 10]
```diff
@@ -0,0 +1,655 @@
...
+
+  jlm::llvm::LlvmRvsdgModule rm(jlm::util::FilePath(""), "", "");
+  auto & rvsdg = rm.Rvsdg();
+
+  auto lambda = jlm::rvsdg::LambdaNode::Create(
+      rvsdg.GetRootRegion(),
+      jlm::llvm::LlvmLambdaOperation::Create(
+          jlm::rvsdg::FunctionType::Create(argTypes, resultTypes),
+          "f",
+          jlm::llvm::Linkage::externalLinkage));
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Just hardcode it instead of using the loops

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 11]
```diff
@@ -0,0 +1,655 @@
...
+
+  // Act & Assert: Verify all argument indices are unique and sequential
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[0]), "a0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[1]), "a1");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[2]), "a2");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[3]), "a3");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[4]), "a4");
+
+  // Note: Lambda results are created when finalize() is called in the function body,
+  // but we can still access subregion()->Results() which will have the correct indices
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: I am not sure what the point of the comment is. Why do you not call finalize and check the port name?

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 12]
```diff
@@ -0,0 +1,655 @@
...
+
+// Test lambda with only memory responses (no register args)
+TEST(BaseHlsTests, TestEmptyLambdaAllMemory)
+{
+  // Arrange: Create a BundleType for memory responses {addr}
+  std::vector<std::pair<std::string, std::shared_ptr<const jlm::rvsdg::Type>>> respElements;
+  respElements.emplace_back("addr", jlm::llvm::PointerType::Create());
+  auto memRespType = std::make_shared<jlm::hls::BundleType>(std::move(respElements));
+
+  jlm::llvm::LlvmRvsdgModule rm(jlm::util::FilePath(""), "", "");
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Could you got through the file and rename the variable name rm to rvsdgModule

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 13]
```diff
@@ -0,0 +1,701 @@
...
+  EXPECT_TRUE(isForbiddenChar('.'));
+  EXPECT_TRUE(isForbiddenChar('!'));
+  EXPECT_TRUE(isForbiddenChar('@'));
+  EXPECT_TRUE(isForbiddenChar(' '));
+  EXPECT_TRUE(isForbiddenChar('\n'));
+  EXPECT_TRUE(isForbiddenChar('\t'));
+}
+
+// Test get_port_name with Input
+TEST(BaseHlsTests, TestGetPortNameInput)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: The test name is misleading: You are are testing the port name of an argument.

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 14]
```diff
@@ -0,0 +1,701 @@
...
+  // Finalize the lambda to create results
+  lambda->finalize({ lambda->GetFunctionArguments()[0], lambda->GetFunctionArguments()[1] });
+
+  // Act & Assert
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[0]), "a0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[1]), "a1");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[2]), "a2");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionResults()[0]), "r0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionResults()[1]), "r1");
+}
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Can those tests not all be somehow merge into a single test?

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 15]
```diff
@@ -0,0 +1,701 @@
...
+
+  // Act
+  auto nodeName1 = baseHls.get_node_name(node1);
+  auto nodeName2 = baseHls.get_node_name(node2);
+
+  // Assert
+  EXPECT_FALSE(nodeName1.empty());
+  EXPECT_FALSE(nodeName2.empty());
+  EXPECT_NE(nodeName1, nodeName2);
+}
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: I think they can be merged into a single test as well.

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 16]
```diff
@@ -0,0 +1,701 @@
...
+          jlm::rvsdg::FunctionType::Create({}, {}),
+          "f",
+          jlm::llvm::Linkage::externalLinkage));
+
+  // Act
+  auto regArgs = TestableBaseHLS().get_reg_args(*lambda);
+
+  // Assert
+  EXPECT_TRUE(regArgs.empty());
+}
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: They can also be merge into a single test

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 17]
```diff
@@ -0,0 +1,701 @@
...
+          jlm::rvsdg::FunctionType::Create({}, {}),
+          "f",
+          jlm::llvm::Linkage::externalLinkage));
+
+  // Act
+  auto regResults = TestableBaseHLS().get_reg_results(*lambda);
+
+  // Assert
+  EXPECT_TRUE(regResults.empty());
+}
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Same here

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 18]
```diff
@@ -0,0 +1,701 @@
...
+
+// Test get_mem_reqs with no memory requests
+TEST(BaseHlsTests, TestGetMemReqsNone)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Here is a test missing for when there is a bundle type.

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 19]
```diff
@@ -0,0 +1,701 @@
...
+
+  // Act
+  auto memReqs = TestableBaseHLS().get_mem_reqs(*lambda);
+
+  // Assert
+  EXPECT_TRUE(memReqs.empty());
+}
+
+// Test get_mem_resps with no memory responses
+TEST(BaseHlsTests, TestGetMemRespsNone)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: same here, test is missing for when there is a bundle type.

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 20]
```diff
@@ -0,0 +1,701 @@
...
+
+  // Act
+  auto memResps = TestableBaseHLS().get_mem_resps(*lambda);
+
+  // Assert
+  EXPECT_TRUE(memResps.empty());
+}
+
+// Test JlmSize with different types
+TEST(BaseHlsTests, TestJlmSize)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Test for state types and bundle types are missing, as well as checking that an exception is thrown if it does not know the type

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 21]
```diff
@@ -0,0 +1,701 @@
...
+
+  // Assert - no forbidden characters should be in the name
+  for (size_t i = 0; i < nodeName.size(); ++i)
+  {
+    EXPECT_FALSE(isForbiddenChar(nodeName[i]));
+  }
+}
+
+// Test port naming with complex types
+TEST(BaseHlsTests, TestPortNamingWithComplexType)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Can be merged with the tests above

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 22]
```diff
@@ -0,0 +1,701 @@
...
+  // Finalize the lambda to create results
+  lambda->finalize({ lambda->GetFunctionArguments()[0] });
+
+  // Act & Assert
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[0]), "a0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionResults()[0]), "r0");
+}
+
+// Test edge cases - empty lambda with only memory responses
+TEST(BaseHlsTests, TestEdgeCaseEmptyLambdaWithMemResp)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Can be merged with tests above

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 23]
```diff
@@ -0,0 +1,701 @@
...
+  auto regArgs = TestableBaseHLS().get_reg_args(*lambda);
+  auto memResps = TestableBaseHLS().get_mem_resps(*lambda);
+
+  // Assert
+  EXPECT_TRUE(regArgs.empty());  // No register args
+  EXPECT_EQ(memResps.size(), 1); // But has memory response
+}
+
+// Test memory request extraction with BundleType results
+TEST(BaseHlsTests, TestMemReqsWithBundleType)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Can be merged with tests above.

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 24]
```diff
@@ -0,0 +1,701 @@
...
+
+  // Act
+  auto memReqs = TestableBaseHLS().get_mem_reqs(*lambda);
+
+  // Assert - no memory requests since there's no BundleType results
+  EXPECT_TRUE(memReqs.empty());
+}
+
+// Test memory response extraction with BundleType arguments
+TEST(BaseHlsTests, TestMemRespsWithBundleType)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Can be merged with the tests above

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 25]
```diff
@@ -0,0 +1,701 @@
...
+  {
+    argTypes.push_back(jlm::rvsdg::BitType::Create(32));
+  }
+
+  auto valueType = jlm::rvsdg::TestType::createValueType();
+  std::vector<std::shared_ptr<const jlm::rvsdg::Type>> resultTypes;
+  for (int i = 0; i < 3; ++i)
+  {
+    resultTypes.push_back(valueType);
+  }
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: just hardcode the types in the lambda

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 26]
```diff
@@ -0,0 +1,701 @@
...
+  // Act
+  auto memResps = TestableBaseHLS().get_mem_resps(*lambda);
+
+  // Assert - should extract the BundleType argument as memory response
+  EXPECT_EQ(memResps.size(), 1);
+  EXPECT_EQ(memResps[0], lambda->GetFunctionArguments()[0]);
+}
+
+// Test port naming with many inputs/outputs to verify index counter behavior
+TEST(BaseHlsTests, TestPortNamingMaxIndices)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: Can also be merged with the tests above

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 27]
```diff
@@ -0,0 +1,701 @@
...
+  // Act & Assert: Verify all argument indices are unique and sequential
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[0]), "a0");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[1]), "a1");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[2]), "a2");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[3]), "a3");
+  EXPECT_EQ(TestableBaseHLS().get_port_name(lambda->GetFunctionArguments()[4]), "a4");
+}
+
+// Test lambda with only memory responses (no register args)
+TEST(BaseHlsTests, TestEmptyLambdaAllMemory)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: can be merged

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 28]
```diff
@@ -0,0 +1,701 @@
...
+  auto regArgs = TestableBaseHLS().get_reg_args(*lambda);
+  auto memResps = TestableBaseHLS().get_mem_resps(*lambda);
+
+  // Assert: No register args, but has memory response
+  EXPECT_TRUE(regArgs.empty());  // No register args
+  EXPECT_EQ(memResps.size(), 1); // But has memory response
+}
+
+// Test lambda with both register and memory arguments
+TEST(BaseHlsTests, TestMixedRegAndMemArgs)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: same here, can be merged

File: jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp [hunk 29]
```diff
@@ -0,0 +1,701 @@
...
+
+  // Assert: One register arg, one memory response
+  EXPECT_EQ(regArgs.size(), 1); // Only 'value' is register arg
+  EXPECT_EQ(regArgs[0], lambda->GetFunctionArguments()[0]);
+  EXPECT_EQ(memResps.size(), 1); // mem_resp is memory response
+  EXPECT_EQ(memResps[0], lambda->GetFunctionArguments()[1]);
+}
+
+// Test multiple BundleType arguments (multiple memory responses)
+TEST(BaseHlsTests, TestMultipleMemResponses)
```
Comment (phate) [inline jlm/hls/backend/rhls2firrtl/BaseHlsTests.cpp]: and again, can be merged

**Review by phate — COMMENTED** [review]
The entire file can be reduced to a hand full of tests. Instead of creating a test for each specific case, just create one test with multiple lambdas with the the different properties. Often, you can just combine them. Most the code that you test just looks through whether it is a bundle type or not.

**Review by phate — APPROVED** [review]
